README.md


<div id="badges"><!-- pkgdown markup -->
<a href="https://CRAN.R-project.org/web/checks/check_results_aroma.apd.html"><img border="0" src="https://www.r-pkg.org/badges/version/aroma.apd" alt="CRAN check status"/></a> <a href="https://github.com/HenrikBengtsson/aroma.apd/actions?query=workflow%3AR-CMD-check"><img border="0" src="https://github.com/HenrikBengtsson/aroma.apd/actions/workflows/R-CMD-check.yaml/badge.svg?branch=develop" alt="R CMD check status"/></a>      
</div>

# aroma.apd: A Probe-Level Data File Format Used by 'aroma.affymetrix' [deprecated] 


## Installation
R package aroma.apd is available on [CRAN](https://cran.r-project.org/package=aroma.apd) and can be installed in R as:
```r
install.packages("aroma.apd")
```


### Pre-release version

To install the pre-release version that is available in Git branch `develop` on GitHub, use:
```r
remotes::install_github("HenrikBengtsson/aroma.apd", ref="develop")
```
This will install the package from source.  

<!-- pkgdown-drop-below -->



NEWS
Package: aroma.apd
==================

Version: 0.6.0-9002 [2022-07-17]
o Fix unnecessary escape of underscores (_) in Rd files.

Version: 0.6.0 [2015-02-24]
o ROBUSTNESS: Added package tests (from examples).
o ROBUSTNESS: Importing suggested functions using ::.
o ROBUSTNESS: Declaring all S3 methods.
o Bumped package dependencies.


Version: 0.5.0 [2014-02-25]
o Bumped up package requirements.


Version: 0.4.1 [2014-01-05]
o Now celToApd() throws a more informative error message if it fails
  to located an APD map file.
o Now the 'aroma.apd' Package object is also available when the
  package is only loaded (but not attached).
o CLEANUP: Removed unnecessary import statements from NAMESPACE.
o ROBUSTNESS: Now writeApd() is atomic, which is achieved by first
  writing to a temporary file which is then renamed upon completion.
o ROBUSTNESS: Some of the examples assumed that the suggested
  'affxparser' was installed and attached.
o BUG FIX: celToApd() assumed that affxparser was already attached.
o Bumped up package version requirements.


Version: 0.4.0 [2013-09-21]
o CLEANUP: Removed fallback attachments of 'R.utils' and 'R.huge' as
  these are no longer needed with R.oo (>= 1.15.1).
o CLEANUP: Package no longer utilizes import(), only importFrom():s.
o CLEANUP: Now only using 'Authors@R' in DESCRIPTION, which is possible
  since R (>= 2.14.0).
o CLEANUP: Dropped obsolete R (< 2.0.0) code that was never used.
o BUG FIX: cdfToApdMap() did not validate and assign argument 'verbose'.
o BUG FIX: Forgot to import trim() from 'R.oo'.


Version: 0.3.0 [2013-08-02]
o CLEANUP: Package no longer attaches/loads R.utils and R.huge (only
  imports them).   However, they will be attached dynamically, when
  certain methods are called.
o Utilizing new startupMessage() of R.oo.
o BUG FIX: Several functions incorrectly assumed that the affxparser
  package had been attached ("loaded").  Although this is almost always
  the case, because this package is only used by aroma.affymetrix which
  loads affxparser, this was an incorrect assumption.


Version: 0.2.5 [2013-05-25]
o Minor speedup by replacing rm() calls with NULL assignments.


Version: 0.2.4 [2013-05-22]
o CRAN POLICY: Now all Rd \usage{} lines are at most 90 characters long.


Version: 0.2.3 [2012-06-30]
o Updated package dependencies.


Version: 0.2.2 [2012-04-04]
o Now package only imports R.methodsS3.


Version: 0.2.1 [2012-04-01]
o CLEANUP: Now package no longer gives an 'R CMD check' NOTE that
  .onAttach() tries to load the 'affxparser' package.


Version: 0.2.0 [2011-07-23]
o Now the package depends on R.huge v0.3.0, which now also
  has a namespace.


Version: 0.1.9 [2011-02-20]
o Now this package will try to install affxparser, if missing
  and in interactive mode.


Version: 0.1.8 [2011-01-31]
o Using packageStartupMessage() instead of cat().


Version: 0.1.7 [2009-10-16]
o Fixed a few broken Rd links to the affxparser package.


Version: 0.1.6 [2009-06-11]
o Made it explicit in the title and the description that this
  package should now be considered deprecated.


Version: 0.1.5 [2009-05-16]
o Updated {read|update|write}Apd() to coerce argument 'writeMap' to
  integer indices.  readApd() also coerce argument 'indices' to the
  same.  Before it used to coerce to doubles, if using R.utils v1.1.4
  or before.  These updates should make no difference to anyone.
  The main purpose is just to use Arguments$getIndices() instead of
  Arguments$getNumerics() where ever possible.
o Fixed a syntax error in one Rdoc comment.
o Renamed the file HISTORY (this one) to NEWS.


Version: 0.1.4 [2007-05-24]
o Added a namespace to the package.


Version: 0.1.3 [2006-06-19]
o Forgot to use autoload("setMethodS3", package="R.oo") etc.
  Packaged would not install on Linux.


Version: 0.1.2 [2006-03-18]
o Added updateApdUnits().
o Now all probe indices are not longer zero-based but one-based.
  However, there is still an option to change this to zero-based
  indices again by using argument 'indexOffset=0'.  This option
  might be remove in the future, because zero-based indices are
  really confusing in R.


Version: 0.1.1 [2006-03-14]
o There is now a quite large set of methods to process APD files.
o All of the methods to re-arrange CDF structures and to create
  cell-index maps have been moved to the affxparser package.


Version: 0.1.0 [2006-02-23]
o Created.



R/old/read.cdf.env.R
read.cdf.env <- function(fname, complementary.logic = TRUE, verbose = 0) {
    requireNamespace("Biobase") || stop("Package not found: Biobase")

    fname <- file.path(dirname(fname), basename(fname))
    if (!file.exists(fname))
        stop(paste("file:", fname, "does not exist."))
    pmmm <- .Call("R_affx_get_pmmm_list",
                  as.character(fname),
                  complementary.logic,
                  verbose = verbose,
                  PACKAGE="affxparser")
    if (is.null(pmmm) || length(pmmm) == 0)
        stop(paste("Error parsing:", fname))
    pmmm <- lapply(pmmm, function(x) {
        tmp <- t(x)
        colnames(tmp) <- c("pm", "mm")
        tmp})
    e <- new.env(hash = TRUE)
    multiassign(names(pmmm), value = pmmm, e)
    return(e)
}

############################################################################
# HISTORY:
# 2006-01-10
# o Added require(Biobase) after removing 'Depends: Biobase' in DESCRIPTION.
# o Added PACKAGE="affxparser" to all .Call() calls. /HB
# o Extracted to its own source file. /HB
############################################################################


R/hide/gtypeCelToPQ.R
########################################################################/**
# @RdocDefault gtypeCelToPQ
#
# @title "Function to immitate Affymetrix' gtype\_cel\_to\_pq software"
#
# \description{
#  @get "title".
# }
#
# @synopsis
#
# \arguments{
#   \item{filename}{The name of a CEL file.}
#   \item{units}{Indices of CDF units to be returned.}
#   \item{...}{Arguments passed to @see "affxparser::readCelUnits".}
#   \item{cdf}{A CDF @list structure, the pathname of the CDF file, or
#     @NULL.  If @NULL, the CDF file corresponding to the chip type of
#     the CEL file is searched for using @see "affxparser::findCdf".}
#   \item{nbrOfQuartets}{The number of probe quartets in the returned
#     @matrix.}
#   \item{verbose}{See @see "R.utils::Verbose".}
# }
#
# \value{
#  Returns an NxK @matrix where N is the number of probesets (SNPs) and
#  K=4*Q where Q is the number of probe quartets (PMA,MMA,PMB,MMB).
#  The rownames corresponds to the probeset names.
# }
#
# @examples "../incl/gtypeCelToPQ.Rex"
#
# \seealso{
#  @see "gtypeCelToPQ".
#  @see "affxparser::applyCdfGroups".
# }
#
# @author
#
# \references{
#  [1] Affymetrix, \emph{Understanding Genotyping Probe Set Structure}, 2005.
#      \url{http://www.affymetrix.com/support/developer/whitepapers/genotyping_probe_set_structure.affx}\cr
# }
#
# @keyword programming
#**/#######################################################################
setMethodS3("gtypeCelToPQ", "default", function(filename, units=NULL, ..., cdf=NULL, nbrOfQuartets=NULL, verbose=FALSE) {
  requireNamespace("affxparser") || throw("Package not loaded: affxparser");


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Validate arguments
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Argument 'filename':
  filename <- Arguments$getReadablePathname(filename, mustExist=TRUE);

  # Argument 'nbrOfQuartets':
  if (!is.null(nbrOfQuartets)) {
    nbrOfQuartets <- Arguments$getInteger(nbrOfQuartets, range=c(1,1024));
  }

  # Argument 'verbose':
  verbose <- Arguments$getVerbose(verbose);


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Get CDF structure
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  if (is.list(cdf)) {
  } else if (is.character(cdf)) {
    cdfFile <- cdf;
  } else {
    verbose && enter(verbose, "Obtaining chip type from CEL header");
    celHeader <- affxparser::readCelHeader(filename);
    chiptype <- celHeader$chiptype;
    verbose && exit(verbose);
    verbose && enter(verbose, "Searching for CDF file");
    cdfFile <- affxparser::findCdf(chiptype);
    verbose && exit(verbose);
  }

  if (!is.list(cdf)) {
    verbose && enter(verbose, "Reading CDF structure");
    cdf <- affxparser::readCdfUnits(cdfFile, units=units, stratifyBy="pmmm");
    verbose && exit(verbose);

    verbose && enter(verbose, "Rearranging CDF structure");
    cdf <- affxparser::applyCdfGroups(cdf, cdfGtypeCelToPQ);
    verbose && exit(verbose);
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Read CEL intensities
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  verbose && enter(verbose, "Reading CEL file");
  cel <- affxparser::readCelUnits(filename, ..., cdf=cdf);
  verbose && exit(verbose);

  cdf <- NULL; # Not needed anymore



  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Create (PMA,MMA,PMB,MMB) matrix
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  verbose && enter(verbose, "Creating matrix");
  # Assume equal number of probes in all probesets, which is also what
  # Affymetrix' gtype_cel_to_pq does.
  if (is.null(nbrOfQuartets)) {
    nbrOfProbes <- length(cel[[1]][[1]][[1]]);
    # Assume a multiple of four probes
    if (nbrOfProbes %% 4 != 0) {
      throw("The first unit does not have a multiple of four (quartet) probes: ", nbrOfProbes);
    }
    nbrOfQuartets <- nbrOfProbes %/% 4;
  }

  nbrOfUnits <- length(cel);
  rownames <- names(cel);
  quartets <- rep(c("pmA", "mmA", "pmB", "mmB"), times=nbrOfQuartets);
  colnames <- paste(quartets, rep(sprintf("%02d", 1:nbrOfQuartets), each=4), sep="");
  dimnames <- list(rownames, colnames);
  res <- matrix(NA, nrow=nbrOfUnits, ncol=4*nbrOfQuartets, dimnames=dimnames);
  cc <- 1:ncol(res);
  for (uu in 1:nbrOfUnits) {
    # ... cel[[uu]][[1]][[1]]
    res[uu,cc] <- .subset2(.subset2(.subset2(cel, uu), 1), 1)[cc];
  }

  verbose && exit(verbose);

  res;
})

############################################################################
# HISTORY:
# 2006-03-28
# o Using .subset2() instead.
# 2006-03-12
# o Added argument 'nbrOfQuartets'.
# 2006-03-11
# o Tried it.  It can be used to generate the same tab-delimited files as
#   Affymetrix' gtype_cel_to_pq application.
# o Created, partly to verify the correctness of affxparser.
############################################################################


R/readApdUnits.R
#########################################################################/**
# @RdocDefault readApdUnits
#
# @title "Reads Affymetrix probe data (APD) as units (probesets)"
#
# @synopsis
#
# \description{
#   @get "title" by using the unit and group definitions in the
#   corresponding Affymetrix CDF file.
#
#   If more than one APD file is read, all files are assumed to be of
#   the same chip type, and have the same read map, if any.
#   It is not possible to read APD files of different types at the
#   same time.
# }
#
# \arguments{
#   \item{filenames}{The filenames of the APD files.  All APD files must
#     be of the same chip type.}
#   \item{units}{An @integer @vector of unit indices specifying which units
#     to be read.  If @NULL, all units are read.}
#   \item{...}{Additional arguments passed to @see "readApd".}
#   \item{transforms}{A @list of exactly \code{length(filenames)}
#     @functions.  If @NULL, no transformation is performed.
#     Values read are passed through the corresponding transform
#     function before being returned.}
#   \item{cdf}{A @character filename of a CDF file, or a CDF @list
#     structure.  If @NULL, the CDF file is searched for by
#     @see "affxparser::findCdf" first starting from the current directory
#     and then from the directory where the first APD file is.}
#   \item{stratifyBy}{Argument passed to low-level method
#     @see "affxparser::readCdfCellIndices".}
#   \item{addDimnames}{If @TRUE, dimension names are added to arrays,
#     otherwise not.  The size of the returned APD structure in bytes
#     increases by 30-40\% with dimension names.}
#   \item{readMap}{A @vector remapping cell indices to file indices.
#     If \code{"byMapType"}, the read map of type according to APD header
#     will be search for and read.  It is much faster to specify the
#     read map explicitly compared with searching for it each time.
#     If @NULL, no map is used.}
#   \item{dropArrayDim}{If @TRUE and only one array is read, the elements of
#     the group field do \emph{not} have an array dimension.}
#   \item{verbose}{See @see "R.utils::Verbose".}
# }
#
# \value{
#   A named @list where the names corresponds to the names of the units
#   read.  Each element of the @list is in turn a @list structure
#   with groups (aka blocks).
# }
#
# \section{Speed}{
#   Since the cell indices are semi-randomized across the array and
#   with units (probesets), it is very unlikely that the read will
#   consist of subsequent cells (which would be faster to read).
#   However, the speed of this method, which uses @see "R.huge::FileVector"
#   to read data, is comparable to the speed of
#   @see "affxparser::readCelUnits", which uses the Fusion SDK
#   (@see "affxparser::readCel") to read data.
# }
#
# @author
#
# @examples "../incl/readApdUnits.Rex"
#
# \seealso{
#   To read CEL units, @see "affxparser::readCelUnits".
#   Internally, the @see "readApd" method is used for read probe data,
#   and @see "readApdMap", if APD file has a map type specified and
#   the read map was not given explicitly.
# }
#
# \references{
#   [1] Affymetrix Inc, Affymetrix GCOS 1.x compatible file formats,
#       June 14, 2005.
#       \url{http://www.affymetrix.com/support/developer/}
# }
#
# @keyword "file"
# @keyword "IO"
#*/#########################################################################
setMethodS3("readApdUnits", "default", function(filenames, units=NULL, ..., transforms=NULL, cdf=NULL, stratifyBy=c("nothing", "pmmm", "pm", "mm"), addDimnames=FALSE, readMap="byMapType", dropArrayDim=TRUE, verbose=FALSE) {
  requireNamespace("affxparser") || throw("Package not loaded: affxparser");


  apdHeader <- NULL;

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Validate arguments
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # Argument 'filenames':
  nbrOfArrays <- length(filenames);
  if (nbrOfArrays == 0)
    throw("Cannot read APD units. No APD files specified.");

  filenames <- file.path(dirname(filenames), basename(filenames));
  missing <- filenames[!file.exists(filenames)];
  if (length(missing)) {
    throw("File(s) not found: ", paste(missing, collapse=", "));
  }

  # Argument 'units' and 'cdf':
  if (is.list(cdf) && !is.null(units)) {
    throw("Arguments 'units' must not be specified if argument 'cdf' is a CDF list structure.");
  }

  # Argument 'units':
  if (is.null(units)) {
  } else if (is.numeric(units)) {
    units <- as.integer(units);
    if (any(units < 1))
      stop("Argument 'units' contains non-positive indices.");
  } else {
    stop("Argument 'units' must be numeric or NULL: ", class(units)[1]);
  }

  # Argument 'cdf':
  searchForCdf <- FALSE;
  if (is.null(cdf)) {
    searchForCdf <- TRUE;
  } else if (is.character(cdf)) {
    cdfFile <- file.path(dirname(cdf), basename(cdf));
    if (!file.exists(cdfFile))
      stop("File not found: ", cdfFile);
    cdf <- NULL;
  } else if (is.list(cdf)) {
    aUnit <- cdf[[1]];
    if (!is.list(aUnit))
      throw("Argument 'cdf' is of unknown format: First unit is not a list.");

    groups <- aUnit$groups;
    if (!is.list(groups))
      throw("Argument 'cdf' is of unknown format: Units Does not contain the list 'groups'.");

    # Check for group fields 'indices' or 'x' & 'y' in one of the groups.
    aGroup <- groups[[1]];

    fields <- names(aGroup);
    if ("indices" %in% fields) {
      cdfType <- "indices";
    } else if (all(c("x", "y") %in% fields)) {
      cdfType <- "x";
      searchForCdf <- TRUE;
    } else {
      throw("Argument 'cdf' is of unknown format: The groups contains neither the fields 'indices' nor ('x' and 'y').");
    }
    aUnit <- groups <- aGroup <- NULL; # Not needed anymore
  } else {
    throw("Argument 'cdf' must be a filename, a CDF list structure or NULL: ", mode(cdf));
  }

  # Argument 'verbose':
  verbose <- Arguments$getVerbose(verbose);

  # Argument 'readMap':
  if (is.null(readMap)) {
    # No probe map specified.
  } else if (identical(readMap, "byMapType")) {
    # Assume that all APD file have the same map type.
    apdHeader <- readApdHeader(filenames[1]);

    mapType <- apdHeader$mapType;
    if (!is.null(mapType)) {
      mapType <- trim(mapType);
      if (is.na(mapType) || nchar(mapType) == 0)
        mapType <- NULL;
    }

    readMap <- mapType;
  }

  if (is.character(readMap)) {
    mapFile <- findApdMap(mapType=readMap);
    if (is.null(mapFile))
      throw("No APD map file found for the given map type: ", readMap);
    readMap <- readApdMap(mapFile)$map;
  }

  # Argument 'dropArrayDim':
  dropArrayDim <- Arguments$getLogical(dropArrayDim);

  # Argument 'addDimnames':
  addDimnames <- Arguments$getLogical(addDimnames);

  # Argument 'transforms':
  if (is.null(transforms)) {
    hasTransforms <- FALSE;
  } else if (is.list(transforms)) {
    if (length(transforms) != nbrOfArrays) {
      throw("Length of argument 'transforms' does not match the number of arrays: ", length(transforms), " != ", nbrOfArrays);
    }
    for (transform in transforms) {
      if (!is.function(transform))
        throw("Argument 'transforms' must be a list of functions.");
    }
    hasTransforms <- TRUE;
  } else {
    throw("Argument 'transforms' must be a list of functions or NULL.");
  }


#  if (length(verbose) != 1)
#    stop("Argument 'verbose' must be a single integer.");
#  verbose <- as.integer(verbose);
#  if (!is.finite(verbose))
#    stop("Argument 'verbose' must be an integer: ", verbose);

  # Argument 'stratifyBy':
  stratifyBy <- match.arg(stratifyBy);


  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # 0. Search for CDF file?
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  if (searchForCdf) {
    verbose && enter(verbose, "Searching for CDF file");

    verbose && enter(verbose, "Reading chip type from first APD file (assuming same chip type for all APD files)");
    if (is.null(apdHeader))
      apdHeader <- readApdHeader(filenames[1]);
    chipType <- apdHeader$chipType;
    verbose && exit(verbose);

    verbose && enter(verbose, "Searching for chip type '", chipType, "'");
    cdfFile <- affxparser::findCdf(chipType=chipType);
    if (length(cdfFile) == 0) {
      # If not found, try also where the first APD file is
      opwd <- getwd();
      on.exit(setwd(opwd));
      setwd(dirname(filenames[1]));
      cdfFile <- affxparser::findCdf(chipType=chipType);
      setwd(opwd);
    }
    verbose && exit(verbose);
    if (length(cdfFile) == 0)
      throw("No CDF file for chip type found: ", chipType);

    verbose && exit(verbose);
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # 1. Read cell indices for units of interest from the CDF file?
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  if (is.null(cdf)) {
    verbose && enter(verbose, "Reading cell indices from CDF file");
    cdf <- affxparser::readCdfCellIndices(cdfFile, units=units, stratifyBy=stratifyBy);
    verbose && exit(verbose);

    # Assume 'cdf' contains only "indices" fields.
    indices <- unlist(cdf, use.names=FALSE);
  } else {
    if (cdfType == "indices") {
      # Clean up CDF list structure from other fields than "indices".
      cdfGetFields <- affxparser::cdfGetFields;
      indices <- affxparser::applyCdfGroups(cdf, cdfGetFields, "indices");
      indices <- unlist(cdf, use.names=FALSE);
    } else {
      verbose && enter(verbose, "Calculating cell indices from (x,y) positions");
      verbose && enter(verbose, "Reading chip layout from CDF file");
      cdfHeader <- affxparser::readCdfHeader(cdfFile);
      verbose && exit(verbose);
      cdfGetFields <- affxparser::cdfGetFields;
      x <- unlist(affxparser::applyCdfGroups(cdf, cdfGetFields, "x"), use.names=FALSE);
      y <- unlist(affxparser::applyCdfGroups(cdf, cdfGetFields, "y"), use.names=FALSE);
      ncol <- cdfHeader$cols;
      indices <- as.integer(y * ncol + x + 1);
      x <- y <- ncol <- NULL; # Not needed anymore
      verbose && exit(verbose);
    }
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # 2. Remapping cell indices?
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  if (!is.null(readMap)) {
    verbose && enter(verbose, "Remapping cell indices (assuming same map for all files)");
    indices <- readMap[indices];
    verbose && exit(verbose);
  }

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # 3. Read signals of the cells of interest from the APD file(s)
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  nbrOfCells <- length(indices);
  nbrOfUnits <- length(cdf);

  verbose && enter(verbose, "Reading ", nbrOfUnits, "*", nbrOfCells/nbrOfUnits, "=", nbrOfCells, " cells from ", nbrOfArrays, " APD files");

  for (kk in seq(length=nbrOfArrays)) {
    filename <- filenames[kk];

    verbose && enter(verbose, "Reading APD data for array #", kk);
    apdTmp <- readApd(filename, indices=indices, readMap=NULL, ..., verbose=verbose, .checkArgs=FALSE);
    verbose && exit(verbose);

    # Will future versions of APD hold more than field? /HB 2006-02-28
    dataFields <- setdiff(names(apdTmp), "header");

    if (kk == 1) {
      verbose && enter(verbose, "Allocating return structure");
      # Allocate the return list structure
      apdTmp$header <- NULL;
      apd <- vector("list", length(apdTmp));
      names(apd) <- names(apdTmp);

      for (name in dataFields) {
        apd[[name]] <- matrix(as.double(0), nrow=nbrOfCells, ncol=nbrOfArrays);
      }

      verbose && exit(verbose);
    }

    for (name in dataFields) {
      # Get read signals
      value <- apdTmp[[name]];
      apdTmp[[name]] <- NULL;

      # Transform signals?
      if (hasTransforms && name %in% dataFields) {
        verbose && enter(verbose, "Transform signals for array #", kk);
        value <- transforms[[kk]](value);
        verbose && exit(verbose);
      }
      apd[[name]][,kk] <- value;
    }
  }
  value <- apdTmp <- NULL; # Not needed anymore
  verbose && exit(verbose);

  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  # 3. Structure APD data in units and groups according to the CDF file
  # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  verbose && enter(verbose, "Structuring data by units and groups");

  # Each group contains the same set of fields
  nbrOfFields <- length(apd);
  fieldNames <- names(apd);
  fields <- vector("list", nbrOfFields);
  names(fields) <- fieldNames;

  # Add a dimension for the arrays, unless only one array is read
  # and the array dimension is not wanted.
  addArrayDim <- (nbrOfArrays >= 2 || !dropArrayDim);

  # For now, assume only one data field.  Extracting the data here, and
  # not in every inner loop, will speed up the whole process about 10 times,
  # e.g. 0.45 seconds instead of 4.3 seconds. /HB 2006-03-04
  data <- apd[[1]];

  seqOfArrays <- list(1:nbrOfArrays);
  offset <- 0;
  res <- lapply(cdf, FUN=function(u) {
    lapply(.subset2(u, "groups"), FUN=function(g) {
      # Same dimensions of all fields
      field <- .subset2(g, 1);  # Faster than g[[1]];
      ncells <- length(field);
      idxs <- offset + 1:ncells;
      offset <<- offset + ncells;

      # Get the target dimension
      dim <- dim(field);
      if (is.null(dim))
        dim <- ncells;

      if (addDimnames) {
        dimnames <- dimnames(field);
        if (is.null(dimnames))
          dimnames <- list(1:ncells);

        # Add an extra dimension for arrays?
        if (addArrayDim) {
          dim <- c(dim, nbrOfArrays);
          dimnames <- c(dimnames, seqOfArrays);
        }

        # Update the data field with dimensions
        # Faster to drop dimensions.
        values <- data[idxs,,drop=TRUE];
        dim(values) <- dim;
        dimnames(values) <- dimnames;
        fields[[1]] <- values;
      } else {
        # Add an extra dimension for arrays?
        if (addArrayDim)
          dim <- c(dim, nbrOfArrays);

        # Update the data field with dimensions
        # Faster to drop dimensions.
        values <- data[idxs,,drop=TRUE];
        if (length(dim) > 1)
          dim(values) <- dim;
        fields[[1]] <- values;
      } # if (addDimnames)

      fields;
    });
  });

  verbose && exit(verbose);

  res;
})


############################################################################
# HISTORY:
# 2006-04-02
# o Using readCdfCellIndices() instead of readCdfUnits().
# 2006-03-28
# o Removed argument 'indexOffset'.
# 2006-03-18
# o Added argument 'indexOffset' and made it one by default (as in R).
# 2006-03-14
# o Added support to obtain the probe map from the map type specified in
#   the APD header.
# 2006-03-04
# o Speed up: By assuming that there is only one single data field in APD
#   files, we can speed up things 10 times.  This is because we do not have
#   to query a list structure in every single inner loop.
# o Removed all gc(). They slow down quite a bit.
# 2006-02-27
# o Created by HB.
############################################################################
